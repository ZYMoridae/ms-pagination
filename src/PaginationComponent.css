.pagination-container {
  margin-top: 24px;
  list-style-type: none;
  text-align: center;
  display: inline-flex;
  padding-left: 0;
}

.pagination-container .pagination-number-wrapper {
  display: inline-flex;
}

.pagination-container .pagination-block {
  position: relative;
  float: left;
  margin-left: -1px;
  width: 35px;
  border: solid 1px #ededed;
  height: 35px;
  color: #3e3e3e;
  text-align: center;
  line-height: 35px;
  cursor: pointer;
  vertical-align: middle;
  z-index: 1;
}

.pagination-container .pagination-block:hover {
  transition: all 0.5s;
  border: 1px solid #ff5000;
  color: #ff5000;
  z-index: 999;
}

.pagination-container .pagination-block.active {
  border: 1px solid #ff5000;
  background-color: #ff5000;
  color: white;
  z-index: 999;
}

.pagination-container .page-number {
  line-height: 19px;
  text-align: center;
}

.pagination-container .page-number .num {
  position: relative;
  z-index: 1;
  display: inline-block;
  width: 100%;
  text-decoration: none;
  font-size: 14px;
  height: 100%;
  line-height: 35px;
  vertical-align: middle;
  font-family: Arial, Helvetica, sans-serif;
}

.pagination-container .page-spread {
  border: none;
}

.pagination-container .page-spread:hover {
  border: none;
}

.pagination-container .previousBtn svg {
  height: 35px;
}

.pagination-container .nextBtn svg {
  height: 35px;
}

.pagination-container .pagination-info {
  position: relative;
  float: left;
  /* margin-left: -1px; */
  /* width: 35px; */
  /* border: solid 1px #ededed; */
  height: 35px;
  color: #8d8d8d;
  text-align: center;
  line-height: 35px;
  /* cursor: pointer; */
  vertical-align: middle;
  z-index: 1;
  margin-left: 15px;
  transition: all 0.5s;
}

.pagination-wrapper {
  width: 100%;
  text-align: center;
  margin-bottom: 32px;
  margin-top: 32px;
}

.pagination-info {
  font-family: sans-serif;
  font-size: 13px;
  color: #8d8d8d;
}

.pagination-info .caption {
  margin-right: 5px;
}

.pagination-wrapper .pagination-info.pagination-meta.mobile {
  display: none;
}

.pagination-container .page-number.mobile {
  display: none;
}
.pagination-container .page-spread.mobile {
  display: none;
}

/* Mobile */
@media (min-width: 320px) and (max-width: 768px) {
  .pagination-container .pagination-info.pagination-meta {
    display: none;
  }

  .pagination-wrapper .pagination-info.pagination-meta.mobile {
    display: block;
  }

  .pagination-wrapper .pagination-info.pagination-meta.mobile .caption {
    margin-left: 5px;
  }

  .pagination-container .page-number.desktop {
    display: none;
  }
  .pagination-container .page-spread.desktop {
    display: none;
  }
  .pagination-container .page-number.mobile {
    display: block;
  }
  .pagination-container .page-spread.mobile {
    display: block;
  }

  .pagination-container .pagination-block {
    position: relative;
    float: left;
    margin-left: -1px;
    width: 28px;
    border: solid 1px #ededed;
    height: 28px;
    color: #3e3e3e;
    text-align: center;
    line-height: 28px;
    cursor: pointer;
    vertical-align: middle;
    z-index: 1;
  }

  .pagination-container .previousBtn svg {
    height: 28px;
  }
  
  .pagination-container .nextBtn svg {
    height: 28px;
  }

  .pagination-container .page-number .num {
    position: relative;
    z-index: 1;
    display: inline-block;
    width: 100%;
    text-decoration: none;
    font-size: 12px;
    height: 100%;
    line-height: 28px;
    vertical-align: middle;
    font-family: Arial, Helvetica, sans-serif;
  }

  .pagination-container .page-spread {
    border: none;
  }
}