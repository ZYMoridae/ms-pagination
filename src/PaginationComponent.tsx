import React from "react";
import "../src/PaginationComponent.css";
import _ from "lodash";
import PaginationComponentProps from "./PaginationComponentProps";
import PaginationParameterMapping from "./PaginationParameterMapping";

import {
  updateUrlParmas,
  getPaginationArray,
  NavigateBeforeRoundedIcon,
  NavigateNextRoundedIcon,
  FirstPageRoundedIcon,
  LastPageRoundedIcon
} from "./PaginationComponentHelper";

type PaginationComponentState = {
  perPage: number;
  orderBy: string;
  page: number;
};

/**
 * Default per page options
 */
const defaultPerPageOptions: Array<number> = [5, 10, 15, 20];

/**
 * Pagination component
 */
class PaginationComponent extends React.Component<
  PaginationComponentProps,
  PaginationComponentState
> {
  constructor(props: any) {
    super(props);
    this.state = {
      perPage: 10,
      orderBy: "",
      page: 1
    };
  }

  componentWillMount() {
    const { initialPage, perPage } = this.props;

    this.setState({
      page: initialPage,
      perPage: perPage
    });
  }

  componentDidMount() {
    const { onPageChanged } = this.props;
    onPageChanged(this.state.page, this.state.perPage, this.state.orderBy);
  }

  pageChanged = (item: number) => {
    this.setState({
      page: item
    });

    const {
      enableUrlParamsUpadte = true,
      paginationParameterMapping = {
        page: "page",
        perPage: "perPage",
        orderBy: "orderBy"
      }
    } = this.props;

    this.props.onPageChanged(item, this.state.perPage, this.state.orderBy);
    if (enableUrlParamsUpadte) {
      updateUrlParmas({
        page: item,
        perPage: this.state.perPage,
        orderBy: this.state.orderBy,
        paginationParameterMapping: paginationParameterMapping
      });
    }
  };

  previousBtnClick = () => {
    let page = this.state.page;

    if (page - 1 >= 1) {
      page = page - 1;
    }

    this.setState({
      page: page
    });

    this.pageChanged(page);
  };

  nextBtnClick = () => {
    let page = this.state.page;

    if (page + 1 <= this.props.count) {
      page = page + 1;
    }

    this.setState({
      page: page
    });

    this.pageChanged(page);
  };

  firstPageBtnClick = () => {
    this.pageChanged(1);
  };

  lastPageBtnClick = () => {
    this.pageChanged(this.props.count);
  };

  perPageOptionOnClickHandler = (event: any) => {
    let _perPage: number = parseInt(event.target.value, 10);
    const {
      enableUrlParamsUpadte = true,
      paginationParameterMapping = {
        page: "page",
        perPage: "perPage",
        orderBy: "orderBy"
      }
    } = this.props;
    this.setState({
      perPage: _perPage
    });
    this.props.onPageChanged(this.state.page, _perPage, this.state.orderBy);
    if (enableUrlParamsUpadte) {
      updateUrlParmas({
        page: this.state.page,
        perPage: _perPage,
        orderBy: this.state.orderBy,
        paginationParameterMapping: paginationParameterMapping
      });
    }
  };

  render() {
    const {
      count,
      nextIcon = <NavigateNextRoundedIcon />,
      previousIcon = <NavigateBeforeRoundedIcon />,
      firstPageNavigationIcon = <FirstPageRoundedIcon />,
      lastPageNavigationIcon = <LastPageRoundedIcon />,
      perPageOptions = defaultPerPageOptions,
      showPerPageOptions = false,
      showFirstPageNavigation = true,
      showLastPageNavigation = true,
      showPaginationMeta = true
    } = this.props;

    return (
      <div className="pagination-wrapper">
        <ul className="pagination-container">
          <div className="pagination-number-block-wrapper">
            {showFirstPageNavigation && (
              <li
                className="pagination-block previousBtn"
                onClick={this.firstPageBtnClick}
              >
                {firstPageNavigationIcon}
              </li>
            )}

            <li
              className="pagination-block previousBtn"
              onClick={this.previousBtnClick}
            >
              {previousIcon}
            </li>

            {/* Desktop */}
            {getPaginationArray(this.state.page, count, false).map(
              (item: any, index: number) =>
                _.isNumber(item) ? (
                  <li
                    className={`pagination-block desktop page-number ${
                      this.state.page === item ? "active" : ""
                    }`}
                    key={index}
                  >
                    <a
                      className="num"
                      onClick={() => {
                        this.pageChanged(item);
                      }}
                    >
                      {item}
                    </a>
                  </li>
                ) : (
                  <li
                    key={index}
                    className="pagination-block desktop page-spread"
                  >
                    ...
                  </li>
                )
            )}

            {/* Mobile */}
            {getPaginationArray(this.state.page, count, true).map(
              (item: any, index: number) =>
                _.isNumber(item) ? (
                  <li
                    className={`pagination-block mobile page-number ${
                      this.state.page === item ? "active" : ""
                    }`}
                    key={index}
                  >
                    <a
                      className="num"
                      onClick={() => {
                        this.pageChanged(item);
                      }}
                    >
                      {item}
                    </a>
                  </li>
                ) : (
                  <li
                    key={index}
                    className="pagination-block mobile page-spread"
                  >
                    ...
                  </li>
                )
            )}

            <li
              className="pagination-block nextBtn"
              onClick={this.nextBtnClick}
            >
              {nextIcon}
            </li>

            {showLastPageNavigation && (
              <li
                className="pagination-block previousBtn"
                onClick={this.lastPageBtnClick}
              >
                {lastPageNavigationIcon}
              </li>
            )}
          </div>

          {showPaginationMeta && (
            <li className="pagination-info pagination-meta">
              <span>Total {count} pages</span>
            </li>
          )}

          {showPerPageOptions && (
            <li className="pagination-info pagination-meta">
              <span className="caption">Per page</span>
              <select
                onChange={this.perPageOptionOnClickHandler}
                value={this.state.perPage}
              >
                {perPageOptions.map((perPageOption, index) => (
                  <option key={index}>{perPageOption}</option>
                ))}
              </select>
            </li>
          )}
        </ul>

        <div className="pagination-info pagination-meta mobile">
          <div>
            <span className="pagination-info">Total {count} pages</span>
            {showPerPageOptions && (
              <span>
                <span className="caption">Per page</span>
                <select
                  onChange={this.perPageOptionOnClickHandler}
                  value={this.state.perPage}
                >
                  {perPageOptions.map((perPageOption, index) => (
                    <option key={index}>{perPageOption}</option>
                  ))}
                </select>
              </span>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default PaginationComponent;
