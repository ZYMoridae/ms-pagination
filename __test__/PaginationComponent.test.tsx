import * as React from "react";
import PaginationComponent from '../src/PaginationComponent';
import { create } from "react-test-renderer";

describe("Feature component", () => {
  test("it matches the snapshot", () => {

    const onPageChanged = (page: number, perPage: number, orderBy: string) => { }

    const component = create(<PaginationComponent onPageChanged={onPageChanged} count={5} initialPage={1} perPage={3}/>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});