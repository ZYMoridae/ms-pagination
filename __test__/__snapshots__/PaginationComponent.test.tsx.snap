// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature component it matches the snapshot 1`] = `
<div
  className="pagination-wrapper"
>
  <ul
    className="pagination-container"
  >
    <div
      className="pagination-number-block-wrapper"
    >
      <li
        className="pagination-block previousBtn"
        onClick={[Function]}
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 0v24H0V0h24z"
            fill="none"
            opacity=".87"
          />
          <path
            d="M17.7 15.89L13.82 12l3.89-3.89c.39-.39.39-1.02 0-1.41-.39-.39-1.02-.39-1.41 0l-4.59 4.59c-.39.39-.39 1.02 0 1.41l4.59 4.59c.39.39 1.02.39 1.41 0 .38-.38.38-1.02-.01-1.4zM7 6c.55 0 1 .45 1 1v10c0 .55-.45 1-1 1s-1-.45-1-1V7c0-.55.45-1 1-1z"
          />
        </svg>
      </li>
      <li
        className="pagination-block previousBtn"
        onClick={[Function]}
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M14.91 6.71c-.39-.39-1.02-.39-1.41 0L8.91 11.3c-.39.39-.39 1.02 0 1.41l4.59 4.59c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L11.03 12l3.88-3.88c.38-.39.38-1.03 0-1.41z"
          />
        </svg>
      </li>
      <li
        className="pagination-block desktop page-number active"
      >
        <a
          className="num"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className="pagination-block desktop page-number "
      >
        <a
          className="num"
          onClick={[Function]}
        >
          2
        </a>
      </li>
      <li
        className="pagination-block desktop page-number "
      >
        <a
          className="num"
          onClick={[Function]}
        >
          3
        </a>
      </li>
      <li
        className="pagination-block desktop page-spread"
      >
        ...
      </li>
      <li
        className="pagination-block mobile page-number active"
      >
        <a
          className="num"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className="pagination-block mobile page-number "
      >
        <a
          className="num"
          onClick={[Function]}
        >
          2
        </a>
      </li>
      <li
        className="pagination-block mobile page-number "
      >
        <a
          className="num"
          onClick={[Function]}
        >
          3
        </a>
      </li>
      <li
        className="pagination-block mobile page-spread"
      >
        ...
      </li>
      <li
        className="pagination-block nextBtn"
        onClick={[Function]}
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M9.31 6.71c-.39.39-.39 1.02 0 1.41L13.19 12l-3.88 3.88c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0l4.59-4.59c.39-.39.39-1.02 0-1.41L10.72 6.7c-.38-.38-1.02-.38-1.41.01z"
          />
        </svg>
      </li>
      <li
        className="pagination-block previousBtn"
        onClick={[Function]}
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
            opacity=".87"
          />
          <path
            d="M6.29 8.11L10.18 12l-3.89 3.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0l4.59-4.59c.39-.39.39-1.02 0-1.41L7.7 6.7c-.39-.39-1.02-.39-1.41 0-.38.39-.38 1.03 0 1.41zM17 6c.55 0 1 .45 1 1v10c0 .55-.45 1-1 1s-1-.45-1-1V7c0-.55.45-1 1-1z"
          />
        </svg>
      </li>
    </div>
    <li
      className="pagination-info pagination-meta"
    >
      <span>
        Total 
        5
         pages
      </span>
    </li>
  </ul>
  <div
    className="pagination-info pagination-meta mobile"
  >
    <div>
      <span
        className="pagination-info"
      >
        Total 
        5
         pages
      </span>
    </div>
  </div>
</div>
`;
